### ISSUE TITLE
CampaignProperties are not thread safe.

### ISSUE THREAD

**Issue by cwisniew** [RPTools/maptool#1247]
## Is your feature request related to a problem? Please describe.
The question was raised for issues #1238, #1239, #1240 can we use JavaFX to recreate these dialogs. Since the <code>CampaignProperties</code> are heavily used in the rendering pipeline, macros, and many other things that are on the Swing EDT <code>CampaignProperties</code> would need to be changed to be thread safe.

### CampaignProperties
This object is not thread safe (this is also the case of many objects it uses).
All collection fields should be final and replaced with the appropriate alternative from java.util.concurrent.
Access to the following fields should protected with a lock.
- defaultSightType
- initiativeOwnerPermissions
- initiativeMovementLock 

### AppPreferences
surprisingly mostly thread safe already, things that are not:	
- setFogOverlayOpacity
- setChatNotificationColor/getChatNotificationColor
- setTrustedPrefixBG/getTrustedPrefixBG
- setTrustedPrefixFG/getTrustedPrefixFG
- addAssetRoot/getAssetRoots/removeAssetRoot
- setMruCampaigns/setMruCampaigns
- setSavedPaintTextures/getSavedPaintTextures

### BarTokenOverlay
This is not thread safe (extends to implementing classes), and this explains some of the wonky behaviour seen in certain situations with them.
I think the easiest/best way to tackle this is to make all implementations of AbstractTokenOverlay immutable and document that **all** implementing classes must be immutable, and when its changed via the gui/macro a new immutable object is created. (may require builder objects to create "like existing").
- DrawnBarTokenOverlay
- MultipleImageBarTokenOverlay
- SingleImageBarTokenOverlay
- TwoImageBarTokenOverlay
- TwoToneBarTokenOverylay
- BooleanTokenOverlay

### SightType
I think the easiest/best way to tackle this is to make SightType immutable and non extendable.

### LookupTable
Not Thread safe, the following fields need to be protected by synchronisation
- name
- defaultRoll
- tableImage
- visible
- allowLookup

<code>LookupEntry</code> should have all of its fields made final.
<code>entryList</code> field should be changed to a </code>CopyOnWriteArrayList</code>

The method <code>getLookup</code> will need some synchronisation when it access any of the above mentioned fields, but the whole method must **not** be synchronised. The line
<code>Result result = expressionParser.evaluate(roll)</code> must be changed so that it **always** runs on the macro execution thread.

### Result
This is not thread safe and must be made thread safe, since all methods are small and quick it should be safe and there is unlikely to be much contention on a Result object it should be safe to use the synchronized keyword on its methods.

### Describe alternatives you've considered
Well the other options are 
1. Don't do it and only forever rely on everything having to run on Swing EDT (too limiting)
2. Make all methods synchronized (very problematic for anything running on Swing EDT).

### Additional context
This not only benefits this dialog, changing things to be thread safe benefits us going forward in many ways.

**Comment by xdy** [comment]
I've started working on this one.

**Comment by xdy** [comment]
What's on my branch might be useful if someone wants to take up this issue so I'll leave the branch alone. I'm unassigning myself from this issue.

**Comment by nmeier** [comment]
While looking at [1264](https://github.com/RPTools/maptool/issues/1264) and [1263](https://github.com/RPTools/maptool/issues/1263) I'm stumbling into a lot of concurrency issues digging deeper into the biz objects.

I'm not sure that adding synchronization all over the place is the right way to go. Synchronization is hard.

Normally I'd expect one thread only to access the Campaign, its zones, properties, etc. at a time. 

**Simpler**: Make everything go through the AWT event thread.

Exceptions: 
Load/save - asynchronous so block the UI when user initiated
Autosave - clone the campaign as done - that should be possible with ConcurrentHashMap (not as is)
Network - push events back to AWT queue with SwingWorker

I see code like this below that uses synchronized incorrectly ([in a constructor](https://stackoverflow.com/questions/23016664/should-i-synchronize-constructor) you don't have to synchronize on property like zone):

https://github.com/RPTools/maptool/blob/fabf84b0692271811e9ff0735ee78679a4c5beb9/src/main

```
/java/net/rptools/maptool/model/Campaign.java#L166

public Campaign(Campaign campaign) {
    name = campaign.getName();
...
    Map<GUID, Zone> zonesToCopy;
    synchronized (zones) {
      zonesToCopy = new LinkedHashMap<>(campaign.zones);
    }

```


Suggestion, make sure MapTool is single threaded where it needs to be by not letting multiple threads update the biz objects in parallel at all.

**Comment by cwisniew** [comment]
> While looking at [1264](https://github.com/RPTools/maptool/issues/1264) and [1263](https://github.com/RPTools/maptool/issues/1263) I'm stumbling into a lot of concurrency issues digging deeper into the biz objects.

There are plenty of issues this is true and I think its going to be a case of addressing them as we find them. 

Several of the lower level ones like Asset are being addressed in an upcoming change... But there will be more... 

> 
> I'm not sure that adding synchronization all over the place is the right way to go. Synchronization is hard.
> 
> Normally I'd expect one thread only to access the Campaign, its zones, properties, etc. at a time.

It would be nice... but consider the case of JavaScript running in a webview for a character sheet, it would mean having to introduce call backs or promises as the request for token information would need to be offloaded into the main thread... So it just ends up pushing the complexity onto people who want to make character sheets or other UI elements in a webview. 


> 
> **Simpler**: Make everything go through the AWT event thread.
> 
> Exceptions:
> Load/save - asynchronous so block the UI when user initiated
> Autosave - clone the campaign as done - that should be possible with ConcurrentHashMap (not as is)

This would be a worthwhile change, but it has implications for campaign persistence for saving/loading from campaign files and backwards compatibility as well as sending campaign information over the network to clients.

> Network - push events back to AWT queue with SwingWorker
> 

People are already clamouring for macros to be moved off the AWT thread (yes the thought scares me...) so I am not sure moving more on to the AWT thread is going to help us in the long run, especially considering webview runs in the JavaFX thread. 

> I see code like this below that uses synchronized incorrectly ([in a constructor](https://stackoverflow.com/questions/23016664/should-i-synchronize-constructor) you don't have to synchronize on property like zone):

This is a bug (see below) the intention is not to syncrhonize the property of the constructed object. 

> 
> [/src/main@`fabf84b`](https://github.com/RPTools/maptool/blob/fabf84b0692271811e9ff0735ee78679a4c5beb9/src/main)
> 
> ```
> /java/net/rptools/maptool/model/Campaign.java#L166
> 
> public Campaign(Campaign campaign) {
>     name = campaign.getName();
> ...
>     Map<GUID, Zone> zonesToCopy;
>     synchronized (zones) {
>       zonesToCopy = new LinkedHashMap<>(campaign.zones);
>     }
> ```
> 

Doh, this should be
```
   synchronized (campaign.zones) {
       zonesToCopy = new LinkedHashMap<>(campaign.zones);
     }
```
So that the synchronized map of zones from the campaign being copied can not be changed while iterating over it (which ```new LinkedHashMap<>(campaign.zones)``` would do. The comment above the code would also make more sense then too :)
It is inadvertently fixing one of the problems that was occuring and making that disappear though (just because there was a synchronized block creating a monitor), but it's definitely synchronizing on the wrong object and should be fixed.

**Comment by nmeier** [comment]
ok, I learned a bit more about the campaign, zone, etc. data structures. It seems doable to have threads update it while performing a 'clone'.
I think a ConcurrentHashMap in those types is the better alternative as you can basically pull out data at any time, anywhere, without risking anything to the map internal structures.
I would apply that down to the Zone, which currently doesn't have any multithreading protection.

I also don't like UI feedback given [right out of the bowls of the cloning logic](https://github.com/RPTools/maptool/blob/03273429b65e33d2c989c2696d85c4e6b0aaad4b/src/main/java/net/rptools/maptool/model/Zone.java#L304)


```
    try {
      grid = (Grid) zone.grid.clone();
      grid.setZone(this);
    } catch (CloneNotSupportedException cnse) {
      MapTool.showError("Trying to copy the zone's grid; no grid assigned", cnse);
    }
```

I can take a crack and this if you guys think this is important enough.

**Comment by Phergus** [comment]
One hiccup with that will be it will break the exporting to previous campaign versions.

**Comment by cwisniew** [comment]
> ok, I learned a bit more about the campaign, zone, etc. data structures. It seems doable to have threads update it while performing a 'clone'.
> I think a ConcurrentHashMap in those types is the better alternative as you can basically pull out data at any time, anywhere, without risking anything to the map internal structures.
> I would apply that down to the Zone, which currently doesn't have any multithreading protection.
> 

As Phergus said, the unfortunate byproduct is that exporting to previous versions will no longer work. It is work opening another issue for it though so we can address it when the campaign format changes greatly (probably the campaign directory issue), and allowing this to be merged.

> I also don't like UI feedback given [right out of the bowls of the cloning logic](https://github.com/RPTools/maptool/blob/03273429b65e33d2c989c2696d85c4e6b0aaad4b/src/main/java/net/rptools/maptool/model/Zone.java#L304)
> 
> ```
>     try {
>       grid = (Grid) zone.grid.clone();
>       grid.setZone(this);
>     } catch (CloneNotSupportedException cnse) {
>       MapTool.showError("Trying to copy the zone's grid; no grid assigned", cnse);
>     }
> ```
> 
> I can take a crack and this if you guys think this is important enough.

Good feedback will always be important :)
